### ISSUE TITLE
`leaky_relu_` wrong gradient with negative slope

### ISSUE THREAD

**Issue by ZisIsNotZis** [pytorch/pytorch#31938]
## 🐛 Bug
When using negative slope in `torch.nn.functional.leaky_relu_`, the back propagated gradient is wrong.

(It turns out this is related to the anomaly detection system)

## To Reproduce
```python
a = torch.tensor([-1., 1.], requires_grad=True)
b = torch.nn.functional.leaky_relu_(a.clone(), -2)
b.backward(torch.ones(2))
a.grad
```
Output is
```
tensor([1., 1.]))
```

## Expected behavior
Output to be
```
tensor([-2., 1.])
```

## Environment
```
Collecting environment information...
PyTorch version: 1.3.1
Is debug build: No
CUDA used to build PyTorch: 10.1.243

OS: Ubuntu Focal Fossa (development branch)
GCC version: (Ubuntu 9.2.1-21ubuntu1) 9.2.1 20191130
CMake version: Could not collect

Python version: 3.7
Is CUDA available: Yes
CUDA runtime version: Could not collect
GPU models and configuration: GPU 0: GeForce GTX 1080
Nvidia driver version: 440.44
cuDNN version: Could not collect

Versions of relevant libraries:
[pip] numpy==1.17.4
[pip] torch==1.3.1
[pip] torchvision==0.4.2
[conda] mkl                       2019.4                      243    https://mirrors.tuna.tsinghua.edu.cn/anaconda/pkgs/main
[conda] pytorch                   1.3.1           py3.7_cuda10.1.243_cudnn7.6.3_0    https://mirrors.tuna.tsinghua.edu.cn/anaconda/cloud/pytorch
[conda] torchvision               0.4.2                py37_cu101    https://mirrors.tuna.tsinghua.edu.cn/anaconda/cloud/pytorch
```

## Additional context
`leaky_relu_` with positive slope is correct. `leaky_relu` (i.e. the out-of-place version) is always correct.

cc @ezyang @gchanan @zou3519 @SsnL @albanD @gqchen

**Comment by XiaobingSuper** [comment]
I don't think this an issue, **a.clone()** will create a new tensor which just have same date with a , not share same memory with a. So after call **a.clone()**, and make inplace operation, **a** 's value will not been changed. you can test it by:
```
a = torch.tensor([-1., 1.], requires_grad=True)
c = a.clone()
b = torch.nn.functional.leaky_relu_(c, -2)
c.register_hook(print)  # will print c.grad
b.backward(torch.ones(2)) # will get c.grad: c.grad will be tensor([1., 1.]): gradx = grady when x>0 or grady* neg_val when x<=0
print(c) # c will be tensor([2., 1.])
print(a) # a still torch.tensor([-1., 1.])
```

**Comment by ZisIsNotZis** [comment]
Oh I totally forgot that. But anyway, since the gradient should be back-propagated back to a, the gradient is wrong in this case.

**Comment by ZisIsNotZis** [comment]
Wait a minute, I think I know why the gradient is "wrong". `leaky_relu_(x)` did not save an extra copy of sign (i.e. `x.sign()`) during forward pass, and it will use the sign of the current `x` during backward pass. But since `x` is now positive, it will assume it was positive.

Should this be considered a bug? Since usually when you do something like this, the version counter should raise some kind of exception.

**Comment by zou3519** [comment]
Mathematically, the gradient is definitely wrong. Autograd's anomaly mode doesn't actually check if the gradient is correct.

Marking this as high-pri because the gradient value is incorrect

**Comment by albanD** [comment]
The issue is that the formula for the gradient (at least in the cpu kernel [here](https://github.com/pytorch/pytorch/blob/99b3f9cac47ba18d24a446d366f63796e18899eb/aten/src/ATen/native/cpu/Activation.cpp#L363-L371)) expects that the sign of the result is the same as the one from the input. Because the slop from the leaky relu is usually positive I guess (not negative as in this case).

The check for the sign should be dependent on the sign of `negval` in this kernel.
Also the corresponding CUDA kernel should be checked.

**Comment by glaringlee** [comment]
Please refer to the PR to see the solution we put there.
